### PR TITLE
Use Ajax call to load organizations on index page

### LINKED ISSUE #504: Use Ajax call to load organizations on index page.
https://github.com/education/classroom/pull/501 closed a #259, :zap: to @dinever 

Even with eager loading the users on the organizations the page load is rather slow.

It would be great if we could load the organizations using ajax so that the user isn't waiting for entire page to load.

/cc @johndbritton for thoughts.

### PR DESCRIPTION
**Pull request by cyhsutw**
Hello guys,

Just implemented to use AJAX calls to load organizations on index page.

Fixes #504.

Demo:

1. For new users without any classrooms available:

none

2. For returning users with some new classrooms (there's a tiny Octocat spinner on the right of the `Classrooms`):

some

It will be appreciated if you could give me some feedback/suggestion. :smiley: 

Thanks!

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -44,6 +44,8 @@
     end
   end
 
+  get '/classrooms-async', to: 'organizations#async_index'
```
Comment (tarebyte) [inline config/routes.rb]: This should be scoped underneath line 34

File: app/controllers/organizations_controller.rb [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
-  skip_before_action :set_organization, :authorize_organization_access, only: [:index, :new, :create]
+  skip_before_action :set_organization, :authorize_organization_access, only: [:index, :new, :create, :async_index]
 
   decorates_assigned :organization
 
   def index
-    @organizations = current_user.organizations.includes(:users).page(params[:page])
+  end
+
+  def async_index
```
Comment (tarebyte) [inline app/controllers/organizations_controller.rb]: I'm not a fan of the name at all tbh.

@johndbritton do you have any other suggestions of what we could call this?

File: config/routes.rb (hunk 1 above)
Comment (cyhsutw) [inline config/routes.rb]: Thanks! I'll fix this with my next commit. :smile:

File: app/controllers/organizations_controller.rb (hunk 2 above)
Comment (cyhsutw) [inline app/controllers/organizations_controller.rb]: Actually I was using `list` but I found it's a little bit confusing (or maybe it's not). 

I would love to hear suggestions from you!

**Comment by johndbritton** [comment]
@tarebyte I'll defer to you on if you think this is the right approach.

File: app/controllers/organizations_controller.rb (hunk 2 above)
Comment (tarebyte) [inline app/controllers/organizations_controller.rb]: @cyhsutw what about `users_organizations`

File: app/controllers/organizations_controller.rb [hunk 3]
```diff
@@ -1,17 +1,22 @@
+# rubocop:disable ClassLength
```
Comment (johndbritton) [inline app/controllers/organizations_controller.rb]: Maybe it's time to start breaking some of this logic out into `lib` with service objects.

File: app/controllers/organizations_controller.rb (hunk 2 above)
Comment (johndbritton) [inline app/controllers/organizations_controller.rb]: I haven't used this kind of pattern before, so I'm sorry I'm not much help with the naming here. Are there any examples that we can look at elsewhere?

Comment (tarebyte) [inline app/controllers/organizations_controller.rb]: So this kind of matches this endpoint

https://github.com/education/classroom/blob/master/app/controllers/stafftools/resources_controller.rb#L8-L12

**Comment by cyhsutw** [comment]
Hi @tarebyte @johndbritton 

Thanks for reviewing the code.

I've renamed the routes and method to `user_organization` and the endpoint is now at `/classrooms/user-organizations`.

Also, I extracted some of the code to `service/organization_service.rb`, not sure if I'm doing it right.

I would love to hear feedback/suggestion from you.

**Comment by nwoodthorpe** [comment]
This PR has been stale for over a year now.

**Pull request closed without merging**